README
NAME
    MooseX::Role::Logger - Provide logging via Log::Any

VERSION
    version 0.003

SYNOPSIS
    In your modules:

        package MyModule;
        use Moose;
        with 'MooseX::Role::Logger';

        sub run {
            my ($self) = @_;
            $self->cry;
        }

        sub cry {
            my ($self) = @_;
            $self->_logger->info("I'm sad");
        }

    In your application:

        use MyModule;
        use Log::Any::Adapter ('File', '/path/to/file.log');

        MyModule->run;

DESCRIPTION
    This role provides universal logging via Log::Any. The class using this
    role doesn't need to know or care about the details of log
    configuration, implementation or destination.

    Use it when you want your module to offer logging capabilities, but
    don't know who is going to use your module or what kind of logging they
    will implement. This role lets you do your part and leaves actual log
    setup and routing to someone else.

    The application that ultimately uses your module can then choose to
    direct log messages somewhere based on its own needs and configuration
    with Log::Any::Adapter.

    This role is based on Moo so it should work with either Moo or Moose
    based classes.

USAGE
  Testing
    Testing with Log::Any is pretty easy, thanks to Log::Any::Test. Just
    load that before Log::Any loads and your log messages get sent to a test
    adapter that includes testing methods:

        use Test::More 0.96;
        use Log::Any::Test;
        use Log::Any qw/$log/;

        use lib 't/lib';
        use MyModule;

        MyModule->new->cry;
        $log->contains_ok( qr/I'm sad/, "got log message" );

        done_testing;

  Customizing
    If you have a whole set of classes that should log with a single
    category, create your own role and set the "_build__logger_category"
    there:

        package MyLibrary::Role::Logger;
        use Moo::Role;
        with 'MooseX::Role::Logger';

        sub _build__logger_category { "MyLibrary" }

    Then in your other classes, use your custom role:

        package MyLibrary::Foo;
        use Moo;
        with 'MyLibrary::Role::Logger'

METHODS
  _logger
    Returns a logging object. See Log::Any for a list of logging methods it
    accepts.

  _build__logger_category
    Override to set the category used for logging. Defaults to the class
    name of the object (which could be a subclass). You can override to lock
    it to a particular name:

        sub _build__logger_category { __PACKAGE__ }

SEE ALSO
    Since MooseX::Role::Logger is universal, you have to use it with one of
    several Log::Any::Adapter classes:

    *   Log::Any::Adapter::File

    *   Log::Any::Adapter::Stderr

    *   Log::Any::Adapter::Stdout

    *   Log::Any::Adapter::ScreenColoredLevel

    *   Log::Any::Adapter::Dispatch

    *   Log::Any::Adapter::Syslog

    *   Log::Any::Adapter::Log4perl

    These other logging roles are specific to particular logging packages,
    rather than being universal:

    *   MooseX::LazyLogDispatch

    *   MooseX::Log::Log4perl

    *   MooseX::LogDispatch

    *   MooseX::Role::LogHandler

    *   MooseX::Role::Loggable (uses Log::Dispatchouli)

    *   Role::Log::Syslog::Fast

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/dagolden/MooseX-Role-Logger/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/MooseX-Role-Logger>

      git clone https://github.com/dagolden/MooseX-Role-Logger.git

AUTHOR
    David Golden <dagolden@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2013 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for MooseX-Role-Logger

0.003     2014-07-03 12:38:11-04:00 America/New_York

    [Changed]

    - Drop minimum required perl to 5.006

0.002     2014-01-24 16:50:08-05:00 America/New_York

    [Incompatible Change]

    - The role now provides private method '_logger' rather than
      public method 'logger' (thanks, mst)

0.001     2013-10-03 00:24:33 America/New_York

    - First release



lib/MooseX/Role/Logger.pm
use strict;
use warnings;

package MooseX::Role::Logger;
# ABSTRACT: Provide logging via Log::Any
our $VERSION = '0.003'; # VERSION

use Moo::Role;
use Types::Standard qw/Str/;

use Log::Any ();

#pod =method _logger
#pod
#pod Returns a logging object.  See L<Log::Any> for a list of logging methods it accepts.
#pod
#pod =cut

has _logger => (
    is       => 'lazy',
    isa      => sub { ref( $_[0] ) =~ /^Log::Any/ }, # XXX too many options
    init_arg => undef,
);

sub _build__logger {
    my ($self) = @_;
    return Log::Any->get_logger( category => $self->_logger_category );
}

has _logger_category => (
    is  => 'lazy',
    isa => Str,
);

#pod =method _build__logger_category
#pod
#pod Override to set the category used for logging.  Defaults to the class name of
#pod the object (which could be a subclass).  You can override to lock it to a
#pod particular name:
#pod
#pod     sub _build__logger_category { __PACKAGE__ }
#pod
#pod =cut

sub _build__logger_category { return ref $_[0] }

1;


# vim: ts=4 sts=4 sw=4 et:

__END__

=pod

=encoding UTF-8

=head1 NAME

MooseX::Role::Logger - Provide logging via Log::Any

=head1 VERSION

version 0.003

=head1 SYNOPSIS

In your modules:

    package MyModule;
    use Moose;
    with 'MooseX::Role::Logger';

    sub run {
        my ($self) = @_;
        $self->cry;
    }

    sub cry {
        my ($self) = @_;
        $self->_logger->info("I'm sad");
    }

In your application:

    use MyModule;
    use Log::Any::Adapter ('File', '/path/to/file.log');

    MyModule->run;

=head1 DESCRIPTION

This role provides universal logging via L<Log::Any>.  The class using this
role doesn't need to know or care about the details of log configuration,
implementation or destination.

Use it when you want your module to offer logging capabilities, but don't know
who is going to use your module or what kind of logging they will implement.
This role lets you do your part and leaves actual log setup and routing to
someone else.

The application that ultimately uses your module can then choose to direct log
messages somewhere based on its own needs and configuration with
L<Log::Any::Adapter>.

This role is based on L<Moo> so it should work with either L<Moo> or L<Moose>
based classes.

=head1 USAGE

=head2 Testing

Testing with L<Log::Any> is pretty easy, thanks to L<Log::Any::Test>.
Just load that before L<Log::Any> loads and your log messages get
sent to a test adapter that includes testing methods:

    use Test::More 0.96;
    use Log::Any::Test;
    use Log::Any qw/$log/;

    use lib 't/lib';
    use MyModule;

    MyModule->new->cry;
    $log->contains_ok( qr/I'm sad/, "got log message" );

    done_testing;

=head2 Customizing

If you have a whole set of classes that should log with a single category,
create your own role and set the C<_build__logger_category> there:

    package MyLibrary::Role::Logger;
    use Moo::Role;
    with 'MooseX::Role::Logger';

    sub _build__logger_category { "MyLibrary" }

Then in your other classes, use your custom role:

    package MyLibrary::Foo;
    use Moo;
    with 'MyLibrary::Role::Logger'

=head1 METHODS

=head2 _logger

Returns a logging object.  See L<Log::Any> for a list of logging methods it accepts.

=head2 _build__logger_category

Override to set the category used for logging.  Defaults to the class name of
the object (which could be a subclass).  You can override to lock it to a
particular name:

    sub _build__logger_category { __PACKAGE__ }

=head1 SEE ALSO

Since MooseX::Role::Logger is universal, you have to use it with one of
several L<Log::Any::Adapter> classes:

=over 4

=item *

L<Log::Any::Adapter::File>

=item *

L<Log::Any::Adapter::Stderr>

=item *

L<Log::Any::Adapter::Stdout>

=item *

L<Log::Any::Adapter::ScreenColoredLevel>

=item *

L<Log::Any::Adapter::Dispatch>

=item *

L<Log::Any::Adapter::Syslog>

=item *

L<Log::Any::Adapter::Log4perl>

=back

These other logging roles are specific to particular logging packages, rather
than being universal:

=over 4

=item *

L<MooseX::LazyLogDispatch>

=item *

L<MooseX::Log::Log4perl>

=item *

L<MooseX::LogDispatch>

=item *

L<MooseX::Role::LogHandler>

=item *

L<MooseX::Role::Loggable> (uses L<Log::Dispatchouli>)

=item *

L<Role::Log::Syslog::Fast>

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/MooseX-Role-Logger/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/MooseX-Role-Logger>

  git clone https://github.com/dagolden/MooseX-Role-Logger.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Provide logging via Log::Any",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "MooseX-Role-Logger",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "NAME" => "MooseX::Role::Logger",
  "PREREQ_PM" => {
    "Log::Any" => 0,
    "Moo::Role" => 0,
    "Types::Standard" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Log::Any::Test" => 0,
    "Moo" => 0,
    "Test::FailWarnings" => 0,
    "Test::More" => "0.96",
    "lib" => 0,
    "version" => 0
  },
  "VERSION" => "0.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "List::Util" => 0,
  "Log::Any" => 0,
  "Log::Any::Test" => 0,
  "Moo" => 0,
  "Moo::Role" => 0,
  "Test::FailWarnings" => 0,
  "Test::More" => "0.96",
  "Types::Standard" => 0,
  "lib" => 0,
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



